Content.Server/Animals/Systems/UdderSystem.cs

```diff
@@ -0,0 +1,153 @@
+                if (udder.Owner.TryGetComponent<HungerComponent>(out var hunger))
+                {
+                    hunger.HungerThresholds.TryGetValue(HungerThreshold.Peckish, out var targetThreshold);
+
+                    // Is there enough nutrition to produce reagent?
+                    if (hunger.CurrentHunger < targetThreshold)
+                        continue;
+                }
+
+                if (!_solutionContainerSystem.TryGetSolution(udder.OwnerUid, udder.TargetSolutionName, out var solution))
+                    continue;
+
+                //TODO: toxins from bloodstream !?
+                _solutionContainerSystem.TryAddReagent(udder.OwnerUid, solution, udder.ReagentId, udder.QuantityPerUpdate, out var accepted);
+                udder.AccumulatedFrameTime = 0;
+            }
+        }
+
+        private void AttemtMilk(EntityUid uid, EntityUid userUid, EntityUid containerUid, UdderComponent udder)
+        {
+            var user = EntityManager.GetEntity(userUid);
+            if (udder.BeingMilked)
+            {
+                udder.Owner.PopupMessage(user, Loc.GetString("udder-system-already-milking"));
+                return;
+            }
+
+            udder.BeingMilked = true;
+
+            var doargs = new DoAfterEventArgs(userUid, 5, default, uid)
+            {
+                BreakOnUserMove = true,
+                BreakOnDamage = true,
+                BreakOnStun = true,
+                BreakOnTargetMove = true,
+                MovementThreshold = 1.0f,
+                TargetFinishedEvent = new MilkingFinishedEvent(userUid, containerUid),
+                TargetCancelledEvent = new MilkingFailEvent()
+            };
+
+            _doAfterSystem.DoAfter(doargs);
+        }
+
+        private void OnMilkingFinished(EntityUid uid, UdderComponent udder, MilkingFinishedEvent ev)
+        {
+            udder.BeingMilked = false;
+
+            if (!_solutionContainerSystem.TryGetSolution(uid, udder.TargetSolutionName, out var solution))
+                return;
+
+            if (!_solutionContainerSystem.TryGetRefillableSolution(ev.ContainerUid, out var targetSolution))
+                return;
+
+            var user = EntityManager.GetEntity(ev.UserUid);
+
+            var quantity = solution.TotalVolume;
+            if(quantity == 0)
+            {
+                udder.Owner.PopupMessage(user, Loc.GetString("udder-system-dry"));
+                return;
+            }
+
+            if (quantity > targetSolution.AvailableVolume)
+                quantity = targetSolution.AvailableVolume;
+
+            var split = _solutionContainerSystem.SplitSolution(uid, solution, quantity);
+            _solutionContainerSystem.TryAddSolution(ev.ContainerUid, targetSolution, split);
+
+            var container = EntityManager.GetEntity(ev.ContainerUid);
+            udder.Owner.PopupMessage(user, Loc.GetString("udder-system-success", ("amount", quantity), ("target", container)));
```

Comment:
same as above, use popupsystem
```suggestion
            var quantity = solution.TotalVolume;
            if(quantity == 0)
            {
            	_popupSystem.PopupEntity(Loc.GetString("udder-system-dry"), uid, Filter.Entities(ev.UserUid));
                return;
            }

            if (quantity > targetSolution.AvailableVolume)
                quantity = targetSolution.AvailableVolume;

            var split = _solutionContainerSystem.SplitSolution(uid, solution, quantity);
            _solutionContainerSystem.TryAddSolution(ev.ContainerUid, targetSolution, split);

            var container = EntityManager.GetEntity(ev.ContainerUid);
            _popupSystem.PopupEntity(Loc.GetString("udder-system-success", ("amount", quantity), ("target", container)), uid, Filter.Entities(ev.UserUid));
```